server/src/main/java/org/elasticsearch/cluster/service/TaskBatcher.java

```diff
@@ -51,42 +52,47 @@ public void submitTasks(List<? extends BatchedTask> tasks, @Nullable TimeValue t
         assert tasks.stream().allMatch(t -> t.batchingKey == firstTask.batchingKey)
             : "tasks submitted in a batch should share the same batching key: " + tasks;
         // convert to an identity map to check for dups based on task identity
+
+        tasksPerBatchingKey.compute(firstTask.batchingKey, (k, existingTasks) -> {
+            assert assertNoDuplicateTasks(tasks, existingTasks);
+            if (existingTasks == null) {
+                return new LinkedHashSet<>(tasks);
+            }
+            existingTasks.addAll(tasks);
```

Comment:
I am not too fond of this pattern. I *think* it works due to how CHM synchronizes on the head of the bucket in both compute and remove. But I think there is no guarantee from CHM to do either. It seems unlikely that compute could run the remapping function on the same key in parallel (since that would break the call only once guarantee), but I am less certain that a future evolution of CHM could not remove the synchronized in remove (do not see a way to do it though).

In short, I think we rely too much on the internals of CHM here in a very central place. Could we perhaps go immutable here and return a list of the two lists - and flatten it when extracted/removed?

Reply:
I think immutable either gets quite tricky (if we keep a list of lists, the timeout task removal gets awkward) or slow if we copy over and over. I'd rather go synchronized I think, it shouldn't be too much overhead here given how we will never have contention on the collection.

Reply:
Though tbh. I'm not sure about the need for this. CHM docs state:

> This class obeys the same functional specification as {@link java.util.Hashtable}
 
which means that remove will always be synchronized with the compute operation I'd assume?

Reply:
It says right after that:

```
This class is fully
 * interoperable with {@code Hashtable} in programs that rely on its
 * thread safety but not on its synchronization details.
```

It is also just nice to have this code "obviously correct", since it is so important here.